circuits/circuits/sig.circom

```diff
@@ -2,26 +2,25 @@ pragma circom 2.1.5;
 
 // TODO: do we need to check that all registers of r,s fit in n bits?
 // Note: The n-bit check is done for pubkey in FlattenPubkey circuit.
-template VerifySignature(n, k) {
+template VerifySignature(hashIns, n, k) {
     signal input r[k];
     signal input s[k];
     signal input msghash[k];
     signal input pubkey[2][k];
 
-    signal input leaf_coins[2];
-    signal input receiver;
+    signal input msg[hashIns];
     signal input signer;
 
     // message hash check
     {
-        signal msghash_computed <== Poseidon(3)(inputs <== [leaf_coins[0], leaf_coins[1], receiver]);
+        signal msghash_computed <== Poseidon(hashIns)(inputs <== msg);
 
         var cumulativeM = 0;
         var nPow = 1;
         for (var i=0; i<k; i++) {
             _ <== Num2Bits(n)(in <== msghash[i]);
             cumulativeM += msghash[i]*nPow;
-            nPow *= n;
+            nPow *= 2**n;
```

Comment:
I just realised there is an attack vector by doing it this way, even if `cumulativeM` never overflows.

If we don't constrain each register in `msg[hashIns]`, an attacker could give them values greater than $2^{64}$ that still add to the correct cumulativeM.

Like, say we're using the prime $7$, and representing a hash as a number between 0 and 3 using two 1 bit field elements. Suppose the hash is $3$, that forces our bits to be $1$ and $1$ ($2^1 + 1 = 3$) if they're constrained to be bits. But if they're unconstrained we can choose `6` and `2`, since $2^{6} + 2 = 66 \equiv 3 \mod 7$. There are many other examples to choose as well.

I prefer decomposing `msg_hash_computed` into registers, but you could probably also fix this with range checks (using `Num2Bits`).

Reply:
i am checking that each element in msghash fits in n bits just above. there can still be a risk that `cumulativeM` overflows but it would be pretty difficult to find a preimage that gives you those suitable values. do you think there's still a risk?

Reply:
Ugh, OK most of my comment was dumb. I did mean `msghash`, and *duh* I missed that you had already constrained it 🤦‍♂️ 

Yeah, after thinking about it for a while I think it's secure. `msghash` has to be validly signed for `ECDSAVerifyNoPubkeyCheck` to pass, and if even `cumulativeM` overflows, this circuit will only accept the overflowed hash anyway. So I don't think it introduces any issues.